### DISCUSSION TITLE
What is the best way to use LocalStorage for credentials with SSR?

### DISCUSSION THREAD

**Discussion by wilsonneto-swe** [vercel/next.js#17628] [Help] [answered]
Hi guys,
we need to use localStorage in this project, and we have a context api that get the credentials from the localStorage and put in his shared state, but I need to this process happens only on the clientside, or be postponed... 

Is there a way to put the useEffect to execute only on the client-side?
Is there a way to execute a callback only when the app is on client-side?

Thanks so much!
Best regards.

**Comment by jamesmosier** [comment, marked as answer]
`useEffect` should not run on the server, but one way to be certain about what code is run in client vs server is to wrap any logic in `typeof window !== 'undefined'`. Next optimizes this code and will remove this code completely from any server bundles.

```js
useEffect(() => {
  if (typeof window !== 'undefined') {
    // do something on the client only!
  }
}, []);
```

**Reply by wilsonneto-swe** [reply]
Great! Thanks! 
But if it is not executed on the server side, will it be executed on the client side? My fear is not to be executed ...

**Reply by jamesmosier** [reply]
`useEffect` is **always** executed on the client side!

**Reply by wilsonneto-swe** [reply]
Thanks so much for this information! Then... It's everything ok with my app... Thanks so much!
